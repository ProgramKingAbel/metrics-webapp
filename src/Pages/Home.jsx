import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import Container from 'react-bootstrap/Container';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Badge from 'react-bootstrap/Badge';
import ListGroup from 'react-bootstrap/ListGroup';
import CloseButton from 'react-bootstrap/CloseButton';
import { fetchStocks } from '../redux/features/stocks/stocksSlice';
import { fetchDetails } from '../redux/features/details/detailsSlice';

const Home = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    dispatch(fetchStocks());
  }, [dispatch]);

  const stocks = useSelector((state) => state.stocks.stocks);

  const [searchQuery, setSearchQuery] = useState('');
  const filteredStocks = stocks
    .filter((stock) => stock.symbol.toLowerCase().includes(searchQuery.toLowerCase()));

  const displayedStocks = searchQuery ? filteredStocks : stocks;

  return (
    <Container>
      <h1 style={{ padding: '1rem 0 0', textAlign: 'center', color: '#21C78F' }}>StoX</h1>
      <Row style={{ margin: '2rem 1rem', alignItems: 'center' }}>
        <Col xs={10} md={10}>
          <Form.Control
            type="text"
            placeholder="Search stock..."
            size="large"
            value={searchQuery}
            onChange={(e) => setSearchQuery(e.target.value)}
          />
        </Col>
        <Col>
          <CloseButton
            aria-label="Clear form"
            onClick={() => setSearchQuery('')}
          />
        </Col>
      </Row>
      <Row
        style={{
          height: '70vh',
          overflowY: 'scroll',
          boxShadow: 'rgba(0, 0, 0, 0.25) 0px 14px 28px, rgba(0, 0, 0, 0.22) 0px 10px 10px',
        }}
      >
        <ListGroup
          horizontal
          style={{
            fontSize: '1rem',
            flexWrap: 'wrap',
          }}
        >
          {
          displayedStocks.map((stock) => (
            <ListGroup.Item
              style={{ width: '200px', height: '200px' }}
              action
              key={stock.symbol}
              onClick={() => {
                dispatch(fetchDetails(stock.symbol));
                navigate(`/details/${stock.symbol}`);
              }}
            >
              <Col><Badge bg="secondary">{ stock.symbol}</Badge></Col>
              <Col style={{ fontWeight: 'bold' }}>{ stock.companyName}</Col>
              <Col>
                <span>
                  Stock Price:
                  {'  '}
                  {parseFloat(stock.price.toFixed(2))}
                </span>
              </Col>
              <Col>
                <span>
                  Volume:
                  {'  '}
                  {stock.volume}
                </span>
              </Col>
              <Col>
                <span>
                  Beta:
                  {'  '}
                  {parseFloat(stock.beta.toFixed(2))}
                </span>
              </Col>
            </ListGroup.Item>
          ))
            }
        </ListGroup>
      </Row>
    </Container>
  );
};

export default Home;
