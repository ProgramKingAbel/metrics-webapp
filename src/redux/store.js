import { configureStore } from '@reduxjs/toolkit';
import stocksReducer from './features/stocks/stocksSlice';
import detailsReducer from './features/details/detailsSlice';

const store = configureStore({
  reducer: {
    stocks: stocksReducer,
    details: detailsReducer,
  },
});

export default store;
