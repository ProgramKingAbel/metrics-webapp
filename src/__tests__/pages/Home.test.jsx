import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import store from '../../redux/store';
import Home from '../../Pages/Home';

test('renders Home component correctly', () => {
  const { container } = render(<Provider store={store}><Router><Home /></Router></Provider>);
  expect(container).toMatchSnapshot();
});
