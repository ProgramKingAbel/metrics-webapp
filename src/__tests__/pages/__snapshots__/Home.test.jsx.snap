// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Home component correctly 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      style="padding: 1rem 0px 0px; text-align: center; color: rgb(33, 199, 143);"
    >
      Stock-Z
    </h1>
    <div
      class="row"
      style="margin: 2rem 1rem; align-items: center;"
    >
      <div
        class="col-md-10 col-10"
      >
        <input
          class="form-control form-control-large"
          placeholder="Search stock..."
          type="text"
          value=""
        />
      </div>
      <div
        class="col"
      >
        <button
          aria-label="Clear form"
          class="btn-close"
          type="button"
        />
      </div>
    </div>
    <div
      class="row"
      style="display: flex; flex-direction: column; margin-left: 1rem; margin-right: 1rem; height: 70vh; overflow-y: scroll; box-shadow: rgba(0, 0, 0, 0.25) 0px 14px 28px, rgba(0, 0, 0, 0.22) 0px 10px 10px;"
    >
      <div
        class="list-group"
        style="font-size: 0.6rem;"
      >
        <div
          aria-disabled="true"
          class="list-group-item disabled"
          style="display: flex;"
          tabindex="-1"
        >
          <div
            bg="secondary"
            class="col"
          >
            symbol
          </div>
          <div
            class="col"
          >
            Company Name
          </div>
          <div
            class="col"
          >
            Price
          </div>
          <div
            class="col"
          >
            Volume
          </div>
          <div
            class="col"
          >
            Beta
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
